crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs

```diff
@@ -678,6 +678,35 @@ impl ReachabilityConstraints {
                 let ty = infer_expression_type(db, test_expr);
                 ty.bool(db).negate_if(!predicate.is_positive)
             }
+            PredicateNode::ReturnsNever(test_expr) => {
+                let ty = infer_expression_type(db, test_expr);
+                if let Type::FunctionLiteral(function_literal) = ty {
+                    let returns_never =
+                        if function_literal
+                            .signature(db)
+                            .overloads
+                            .iter()
+                            .all(|overload| {
+                                // HACK: for now, require that *all* overloads are annotated with
+                                // returning `Never`
+                                // Ideally, if only some overloads return `Never`, we should consider
+                                // the types of the arguments.
+                                overload.return_ty.is_some_and(|return_type| {
+                                    return_type.is_equivalent_to(db, Type::Never)
+                                })
+                            })
+                        {
+                            Truthiness::AlwaysTrue
+                        } else {
+                            Truthiness::AlwaysFalse
+                        };
+                    returns_never.negate_if(!predicate.is_positive)
+                } else {
+                    // Should I add a panic here?
+                    // What about methods / other callables which are not functions?
```

Comment:
I think we should at the very least support `Type::Callable` also. Ideally we'd support callable instances, bound methods, etc as well. I think the way to do this would be adding a `Type::into_callable_type -> Option<Type>` that always returns `Some(Type::Callable(...))` or `None` -- most of the pieces needed for that are already in place. For clarity it could be added in a separate PR (either before or after this one).

Reply:
This now uses the new `into_callable` which was added in https://github.com/astral-sh/ruff/pull/19130.